src/main/java/hudson/matrix/MatrixBuild.java

```diff
@@ -350,15 +350,26 @@ protected Result doRun(BuildListener listener) throws Exception {
             MatrixProject p = getProject();
             PrintStream logger = listener.getLogger();
 
-            // give axes a chance to rebuild themselves
-            activeConfigurations = p.rebuildConfigurations(this);
+            try {
+                // lock the project
+                p.buildLock();
```

Comment:
- The lock should not be used from the outside. The logic of `listUpAggregators` belongs to `MatrixProject` where the locking should take place.
- What operation can actually race with this? I would say it is configuration update and perhaps rebuilding configurations but I do not see them synchronized over the same lock.